packages/editor/src/lib/editor/managers/SnapManager/HandleSnaps.ts

```diff
@@ -52,23 +61,45 @@ export class HandleSnaps {
 	}): SnapData | null {
 		const snapThreshold = this.manager.getSnapThreshold()
 
-		// Find the nearest point that is within the snap threshold
-		let minDistance = snapThreshold
-		let nearestPoint: Vec | null = null
+		// We snap to two different parts of the shape's handle snap geometry:
+		// 1. The `outline`. This describes the outline of the shape, and we just snap to the
+		//    nearest point on that outline.
+		// 2. The `points`. These are handles or other key points that we want to snap to with a
+		//    higher priority than the normal outline snapping.
+		let minDistanceForOutline = snapThreshold
+		let nearestPointOnOutline: Vec | null = null
+
+		let minDistanceForSnapPoint = snapThreshold
+		let nearestSnapPoint: Vec | null = null
 
 		for (const shapeId of this.manager.getSnappableShapes()) {
-			const handleSnapOutline = this.getSnapGeometryCache().get(shapeId)?.outline
-			if (!handleSnapOutline) continue
+			const snapGeometry = this.getSnapGeometryCache().get(shapeId)
+			if (!snapGeometry) continue
 
 			const shapePageTransform = assertExists(this.editor.getShapePageTransform(shapeId))
 			const pointInShapeSpace = this.editor.getPointInShapeSpace(shapeId, handlePoint)
-			const nearestShapePointInShapeSpace = handleSnapOutline.nearestPoint(pointInShapeSpace)
-			const nearestInPageSpace = shapePageTransform.applyToPoint(nearestShapePointInShapeSpace)
-			const distance = Vec.Dist(handlePoint, nearestInPageSpace)
 
-			if (distance < minDistance) {
-				minDistance = distance
-				nearestPoint = nearestInPageSpace
+			if (snapGeometry.outline) {
+				const nearestShapePointInShapeSpace = snapGeometry.outline.nearestPoint(pointInShapeSpace)
+				const nearestInPageSpace = shapePageTransform.applyToPoint(nearestShapePointInShapeSpace)
+				const distance = Vec.Dist(handlePoint, nearestInPageSpace)
+
+				if (distance < minDistanceForOutline) {
+					minDistanceForOutline = distance
+					nearestPointOnOutline = nearestInPageSpace
+				}
+			}
+
+			if (snapGeometry.points) {
+				for (const snapPointInShapeSpace of snapGeometry.points) {
+					const snapPointInPageSpace = shapePageTransform.applyToPoint(snapPointInShapeSpace)
+					const distance = Vec.Dist(handlePoint, snapPointInPageSpace)
+
+					if (distance < minDistanceForSnapPoint) {
+						minDistanceForSnapPoint = distance
+						nearestSnapPoint = snapPointInPageSpace
+					}
+				}
 			}
 		}
```

Comment:
Nonblocking nitpick:

Snap point hits always take precedence over outline hits, regardless of which is closer. So i think you could do the snap point checking first, then if a hit was found you can avoid doing the outline checking (including the `additionalSegments` bit below).

Reply:
yeah, to make this clearer, what do you think of something like this?
```ts
const pointsToTest = [...snapGeometry.points ?? []]
for (const shapeId of this.manager.getSnappableShapes()) {
	...
	if (snapGeometry.outline) {
		const nearestShapePointInShapeSpace = snapGeometry.outline.nearestPoint(pointInShapeSpace)
		pointsToTest.push(nearestShapePointInShapeSpace)
	}
}
for (const segment of additionalSegments) {
	const nearestOnSegment = Vec.NearestPointOnLineSegment(segment[0], segment[1], handlePoint)
	pointsToTest.push(nearestOnSegment)
}
for (const point of pointsToTest) {
	const snapPointInPageSpace = shapePageTransform.applyToPoint(point)
	const distance = Vec.Dist(handlePoint, snapPointInPageSpace)

	if (distance < minDistanceForSnapPoint) {
		minDistanceForSnapPoint = distance
		nearestSnapPoint = snapPointInPageSpace
	}
}
```

related to this, just checking: does the additionalSegments portion need to also do `shapePageTransform.applyToPoint`?

Reply:
good idea! reworked this a bit

> related to this, just checking: does the additionalSegments portion need to also do shapePageTransform.applyToPoint?
nope, these come in in page space